staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -242,3 +248,54 @@ func generateKey(length int) (key []byte, err error) {
 
 	return key, nil
 }
+
+func validateEncryptedObject(o *kmstypes.EncryptedObject) error {
+	if o == nil {
+		return fmt.Errorf("encrypted object is nil")
+	}
+	if o.EncryptedDEK == nil {
+		return fmt.Errorf("encrypted dek is nil")
+	}
+	if o.EncryptedData == nil {
+		return fmt.Errorf("encrypted data is nil")
+	}
+	if err := validateKeyID(o.KeyID); err != nil {
+		return fmt.Errorf("failed to validate key id: %w", err)
+	}
+	if err := validateAnnotations(o.Annotations); err != nil {
+		return fmt.Errorf("failed to validate annotations: %w", err)
+	}
+	return nil
+}
+
+// validateAnnotations tests the following:
+//  1. The annotations key is what Kubernetes calls a "qualified name".
+//     This validation is the same as the label key name validation.
+//  2. The size of annotations keys + values is less than 32 kB.
+func validateAnnotations(annotations map[string][]byte) error {
+	var errs []error
+	var totalSize uint64
+	for k, v := range annotations {
+		for _, msg := range validation.IsQualifiedName(k) {
+			errs = append(errs, fmt.Errorf("invalid annotation key %q: %s", k, msg))
+		}
+		totalSize += uint64(len(k)) + uint64(len(v))
+	}
+	if totalSize > annotationsMaxSize {
+		errs = append(errs, fmt.Errorf("total size of annotations is %d, which exceeds the max size of %d", totalSize, annotationsMaxSize))
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+// validateKeyID tests the following:
+// 1. The keyID is not empty.
+// 2. The size of keyID is less than 1 kB.
+func validateKeyID(keyID string) error {
+	if keyID == "" {
```

Comment:
nit: kube convention is to use `len`

```suggestion
	if len(keyID) == 0 {
```